ovn_k8s/modes/overlay.py

```diff
@@ -438,3 +467,332 @@ def sync_services(self, services):
+        # and ensure the same rule is not created if already defined in the
+        # subsequent step. The current code will leave a pod without security
+        # rules or with incomplete security rules if an error occurs.
+        self.remove_acls(pod_id, remove_default=True)
+        # TODO(salv-orlando): lswitch & lport can be easily cached
+        lport_name = self._build_logical_port_name(pod_ns, pod_name)
+        lport_uuid = ovn_nbctl("--data=bare", "--no-heading",
+                               "--columns=_uuid", "find",
+                               "logical_switch_port",
+                               "name=%s" % lport_name)
+        lswitch_name = ovn_nbctl("--data=bare", "--no-heading",
+                                 "--columns=name", "find",
+                                 "logical_switch",
+                                 "ports{>=}%s" % lport_uuid)
+        for pseudo_acl in pseudo_acls:
+            match = self._create_acl_match(lport_name, pseudo_acl)
+            vlog.dbg("Pod: %s (Namespace:%s): ACL match: %s" % (
+                pod_name, pod_ns, match))
+            ovn_acl_data = (pseudo_acl[0], match, pseudo_acl[3])
+            # TODO(salv-orlando): Also store policy name in external ids.
+            # It could be useful for debugging
+            self._create_acl(pod_id, lswitch_name, lport_name, *ovn_acl_data)
+
+    def _build_rule_address_set_name(self, policy_name, namespace, rule_data):
+        hasher = hashlib.md5()
+        hash_data = json.dumps(
+            rule_data.get('ports'), ALL_PORTS).encode('utf-8')
+        hasher.update(hash_data)
+        rule_hash = hasher.hexdigest()
+        return "%s_%s_%s" % (policy_name, namespace, rule_hash)
+
+    def create_address_set(self, namespace, policy_data):
+        for rule in policy_data['spec']['ingress']:
+            address_set = self._build_rule_address_set_name(
+                policy_data['metadata']['name'],
+                namespace,
+                rule)
+            # First verify that the address set does not already exist
+            if not ovn_nbctl("--data=bare", "--no-heading",
+                             "--columns=_uuid", "find",
+                             "address_set", "name=%s" % address_set):
+                ovn_nbctl('create', 'address_set', 'name=%s' % address_set)
+
+    def destroy_address_set(self, namespace, policy_data):
+        for rule in policy_data['spec']['ingress']:
+            address_set = self._build_rule_address_set_name(
+                policy_data['metadata']['name'],
+                namespace,
+                rule)
+            ovn_nbctl('destroy', 'address_set', address_set)
+
+    def add_to_address_set(self, pod_ip, namespace, policy_data):
+        for rule in policy_data['spec']['ingress']:
+            rule_address_set = self._build_rule_address_set_name(
+                policy_data['metadata']['name'],
+                namespace,
+                rule)
```

Comment:
I thought I added this check, but as I refactored this code many time I might have removed it.